### PR TITLE
[G-API] Python ROI generic inference

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Build configuration
```
Xforce_builders_only=linux,docs
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.1.0:20.04
build_image:Custom Win=openvino-2021.1.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@dmatveev  @smirnov-alexey Could you take a look ?

**Comment by smirnov-alexey** [comment]
@TolyaTalamanov seems like it comes atop of a bunch of other PRs. Can't you separate the changes here?

File: modules/python/src2/cv2.cpp [hunk 1]
```diff
@@ -2185,7 +2185,6 @@ static int convert_to_char(PyObject *o, char *dst, const ArgInfo& info)
 #include "pyopencv_generated_types_content.h"
 #include "pyopencv_generated_funcs.h"
 
-
```
Comment (alalek) [inline modules/python/src2/cv2.cpp:2188]: Please avoid unnecessary changes

Comment (TolyaTalamanov) [reply modules/python/src2/cv2.cpp]: Done

**Comment by TolyaTalamanov** [comment]
@smirnov-alexey Could you have a look, please ?

**Comment by TolyaTalamanov** [comment]
@AsyaPronina @smirnov-alexey Could you have a look, please ?

File: modules/gapi/misc/python/test/test_gapi_infer.py [hunk 2]
```diff
@@ -9,23 +9,205 @@
...
+        net.setPreferableTarget(cv.dnn.DNN_TARGET_CPU)
+
+        blob = cv.dnn.blobFromImage(img)
+
+        net.setInput(blob)
+        return net.forward(net.getUnconnectedOutLayersNames())
+
+
+    def make_roi(self, img, roi):
+        return img[roi[1]:roi[1] + roi[3], roi[0]:roi[0] + roi[2], ...]
```
Comment (smirnov-alexey) [inline modules/gapi/misc/python/test/test_gapi_infer.py:24]: How does this work?

File: modules/gapi/misc/python/test/test_gapi_infer.py [hunk 3]
```diff
@@ -9,23 +9,205 @@
...
 
 
     def test_age_gender_infer(self):
+        # NB: Check IE
+        if not cv.dnn.DNN_TARGET_CPU in cv.dnn.getAvailableTargets(cv.dnn.DNN_BACKEND_INFERENCE_ENGINE):
+            return
+
+        root_path    = '/omz_intel_models/intel/age-gender-recognition-retail-0013/FP32/age-gender-recognition-retail-0013'
```
Comment (smirnov-alexey) [inline modules/gapi/misc/python/test/test_gapi_infer.py:32]: Don't we have some policy regarding our tests, that if we don't find models/data, we don't run the tests? Won't those tests start failing later with new paths?

File: modules/gapi/misc/python/pyopencv_gapi.hpp [hunk 4]
```diff
@@ -318,12 +320,9 @@ static cv::GRunArg extract_run_arg(const cv::GTypeInfo& info, PyObject* item)
                     reinterpret_cast<pyopencv_gapi_wip_IStreamSource_t*>(item)->v;
                 return source;
             }
-            else
```
Comment (smirnov-alexey) [inline modules/gapi/misc/python/pyopencv_gapi.hpp:321]: I would keep the `else` here, but it's up to you

**Review by smirnov-alexey — APPROVED** [review]

**Comment by TolyaTalamanov** [comment]
> @TolyaTalamanov seems like it comes atop of a bunch of other PRs. Can't you separate the changes here?

This bunch of PRs already in master, this only actual separate changes

File: modules/gapi/misc/python/test/test_gapi_infer.py (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/misc/python/test/test_gapi_infer.py:24]: roi - is a tuple (x,y,h,w)

File: modules/gapi/misc/python/test/test_gapi_infer.py (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/misc/python/test/test_gapi_infer.py:32]: I believe we don't care about it in these tests.

File: modules/gapi/include/opencv2/gapi/infer.hpp [hunk 5]
```diff
@@ -636,11 +636,38 @@ infer2(const std::string& tag,
     return cv::GInferListOutputs{std::move(call)};
 }
 
+// NB: Functions defined below used for python.
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/infer.hpp:639]: are used for Python bindings.

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/infer.hpp:639]: ...but does it make to move it to another namespace then?

Also, does it make sense to define it here or in the python bindings header?

File: modules/gapi/misc/python/shadow_gapi.hpp [hunk 6]
```diff
@@ -27,13 +27,28 @@ namespace cv
        GAPI_WRAP void setInput(const std::string& name, const cv::GFrame& value);
    };
 
+   class GAPI_EXPORTS_W_SIMPLE GInferListInputs
+   {
+   public:
+       GAPI_WRAP GInferListInputs();
+       GAPI_WRAP void setInput(const std::string& name, const cv::GArray<cv::GMat>& value);
+       GAPI_WRAP void setInput(const std::string& name, const cv::GArray<cv::Rect>& value);
```
Comment (dmatveev) [inline modules/gapi/misc/python/shadow_gapi.hpp:35]: where's the implementation?

Comment (dmatveev) [reply modules/gapi/misc/python/shadow_gapi.hpp:35]: I only see .hpp files in this MR

File: modules/gapi/include/opencv2/gapi/infer.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer.hpp:639]: > ...but does it make to move it to another namespace then?
> 
> Also, does it make sense to define it here or in the python bindings header?

Make sense to move it in the python bindings header, but not change namespace, otherwise it'll reflected on python API:
```
cv.gapi.python.infer(...) - looks ugly
cv.gapi.infer(...) - looks ok
```

File: modules/gapi/misc/python/shadow_gapi.hpp (hunk 6 above)
Comment (TolyaTalamanov) [reply modules/gapi/misc/python/shadow_gapi.hpp:35]: There is no implementation, this file isn't compiled at all.

Comment (TolyaTalamanov) [reply modules/gapi/misc/python/shadow_gapi.hpp:35]: It's just a trigger for bindings generator

File: modules/gapi/include/opencv2/gapi/infer.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer.hpp]: Done

File: modules/gapi/misc/python/shadow_gapi.hpp (hunk 6 above)
Comment (dmatveev) [reply modules/gapi/misc/python/shadow_gapi.hpp:35]: But why do we need it then? How it works? Can you please comment it in the code so then strangers like me wouldn't ask those questions again and again?

**Review by dmatveev — APPROVED** [review]
Please put the comment to the python wrapper thing somewhere later.

**Comment by TolyaTalamanov** [comment]
@alalek Can it be merged ?

File: modules/gapi/misc/python/test/test_gapi_infer.py (hunk 3 above)
Comment (alalek) [inline modules/gapi/misc/python/test/test_gapi_infer.py:32]: Remove leading slash `/`

File: modules/gapi/misc/python/test/test_gapi_infer.py (hunk 2 above)
Comment (alalek) [reply modules/gapi/misc/python/test/test_gapi_infer.py:24]: > (x,y,h,w)

Probably it is `(x,y,w,h)`

Clear code (to avoid unclear error messages):

```
x, y, w, h = roi
return img[y:y+h, x:x+w, ...]
```

File: modules/gapi/misc/python/test/test_gapi_core.py [hunk 7]
```diff
@@ -13,7 +13,7 @@
           ('cpu'    , cv.gapi.core.cpu.kernels()),
           ('fluid'  , cv.gapi.core.fluid.kernels())
           # ('plaidml', cv.gapi.core.plaidml.kernels())
-      ]
+       ]
```
Comment (alalek) [inline modules/gapi/misc/python/test/test_gapi_core.py:16]: Perhaps whole block should be reformatted
